runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/WorkerCustomSources.java

```diff
@@ -247,18 +247,37 @@ private static <T> SourceOperationResponse performSplitTyped(
           serializedSize);
     }
 
+    List<BoundedSource<T>> bundlesBeforeCoalesce = bundles;
     int numBundlesBeforeRebundling = bundles.size();
     // To further reduce size of the response and service-side memory usage, coalesce
     // the sources into numBundlesLimit compressed serialized bundles.
-    if (bundles.size() > numBundlesLimit) {
+    while (serializedSize > apiByteLimit || bundles.size() > numBundlesLimit) {
+      // bundle size constrained by API limit, adds 5% allowance
+      int targetBundleSizeApiLimit = (int) (bundles.size() * apiByteLimit / serializedSize * 0.95);
+      // bundle size constrained by numBundlesLimit
+      int targetBundleSizeBundleLimit = Math.min(numBundlesLimit, bundles.size() - 1);
+      int targetBundleSize = Math.min(targetBundleSizeApiLimit, targetBundleSizeBundleLimit);
+
+      if (targetBundleSize <= 1) {
+        String message =
+            String.format(
+                "Unable to coalesce the sources into compressed serialized bundles to satisfy the "
+                    + "allowable limit when splitting %s. With %d bundles, total serialized size "
+                    + "of %d bytes is still larger than the limit %d. For more information, please "
+                    + "check the corresponding FAQ entry at "
+                    + "https://cloud.google.com/dataflow/pipelines/troubleshooting-your-pipeline",
+                source, bundles.size(), serializedSize, apiByteLimit);
```

Comment:
Shall we suggest this workaround with `withTemplateCompatiblity`?

Reply:
This option is BigQueryIO.read only while the bug is generic to dataflow worker. With that option the read teansform will expand differently and likely avoided the problematic code path here (haven't verify)

Reply:
SG

Reply:
The workaround will be using Runner v2. That will convert sources into SDFs that do not run into this source split limit. We should suggest that here.

Reply:
Hi @chamikaramj is this code path only for Dataflow runner v1? If so we could document the recommendation of runner v2 also in https://cloud.google.com/dataflow/docs/guides/common-errors#boundedsource-objects-splitintobundles

Reply:
Yeah, this codepath (source split operation) is only get hit by Runner v1. +1 for documenting in the Website.